# Feature Engineering
The goal of this stage is to create meaningful features that improve model performance and better capture customer behavior patterns.

Feature engineering transforms raw data into informative variables that help machine learning models identify churn risk more effectively.

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

from IPython.core.display import HTML

HTML("""
<style>
.dataframe table, .dataframe th, .dataframe td { font-size: 12px; }
div.output_scroll { overflow-x: auto; }
</style>
""")

In [2]:
data = pd.read_csv("telco_cleaned.csv")

data.head()

,gender,age,under_30,senior_citizen,married,dependents,number_of_dependents,city,zip_code,latitude,longitude,population,referred_a_friend,number_of_referrals,tenure_in_months,offer,phone_service,avg_monthly_long_distance_charges,multiple_lines,internet_service,internet_type,avg_monthly_gb_download,online_security,online_backup,device_protection_plan,premium_tech_support,streaming_tv,streaming_movies,streaming_music,unlimited_data,contract,paperless_billing,payment_method,monthly_charge,total_charges,total_refunds,total_extra_data_charges,total_long_distance_charges,total_revenue,satisfaction_score,churn_label
0,Male,78,No,Yes,No,No,0,Los Angeles,90022,34.023810,-118.156582,68701,No,0,1,No Offer,No,0.00,No,Yes,DSL,8,No,No,Yes,No,No,Yes,No,No,Month-to-Month,Yes,Bank Withdrawal,39.65,39.65,0.00,20,0.00,59.65,3,Yes
1,Female,74,No,Yes,Yes,Yes,1,Los Angeles,90063,34.044271,-118.185237,55668,Yes,1,8,Offer E,Yes,48.85,Yes,Yes,Fiber Optic,17,No,Yes,No,No,No,No,No,Yes,Month-to-Month,Yes,Credit Card,80.65,633.30,0.00,0,390.80,1024.10,3,Yes
2,Male,71,No,Yes,No,Yes,3,Los Angeles,90065,34.108833,-118.229715,47534,No,0,18,Offer D,Yes,11.33,Yes,Yes,Fiber Optic,52,No,No,No,No,Yes,Yes,Yes,Yes,Month-to-Month,Yes,Bank Withdrawal,95.45,1752.55,45.61,0,203.94,1910.88,2,Yes
3,Female,78,No,Yes,Yes,Yes,1,Inglewood,90303,33.936291,-118.332639,27778,Yes,1,25,Offer C,Yes,19.76,No,Yes,Fiber Optic,12,No,Yes,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,98.50,2514.50,13.43,0,494.00,2995.07,2,Yes
4,Female,80,No,Yes,Yes,Yes,1,Whittier,90602,33.972119,-118.020188,26265,Yes,1,37,Offer C,Yes,6.33,Yes,Yes,Fiber Optic,14,No,No,No,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,76.50,2868.15,0.00,0,234.21,3102.36,2,Yes


In [3]:
data.shape

(7043, 41)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 41 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             7043 non-null   object 
 1   age                                7043 non-null   int64  
 2   under_30                           7043 non-null   object 
 3   senior_citizen                     7043 non-null   object 
 4   married                            7043 non-null   object 
 5   dependents                         7043 non-null   object 
 6   number_of_dependents               7043 non-null   int64  
 7   city                               7043 non-null   object 
 8   zip_code                           7043 non-null   int64  
 9   latitude                           7043 non-null   float64
 10  longitude                          7043 non-null   float64
 11  population                         7043 non-null   int64

In [5]:
data.isna().sum().sum()

np.int64(0)

### 🔍 Data Quality Check

- The dataset contains 7043 rows and 41 columns after cleaning.
- No missing values are present.
- Most features are categorical (object type), which confirms the need for encoding.
- Data is now ready for feature engineering.

### 📌 Key Observation

A large portion of features are categorical (Yes/No or multi-class).
This confirms that encoding will be a major step in feature engineering.

In [6]:
data["gender"] = data["gender"].map({
    "Male": 1,
    "Female": 0
})
data["gender"].value_counts()

gender
1    3555
0    3488
Name: count, dtype: int64

In [7]:
data["gender"].isna().sum()

np.int64(0)

In [8]:
binary_cols = [
    'senior_citizen',
    'married',
    'dependents',
    'referred_a_friend',
    'phone_service',
    'multiple_lines',
    'internet_service',
    'online_security',
    'online_backup',
    'device_protection_plan',
    'premium_tech_support',
    'streaming_tv',
    'streaming_movies',
    'streaming_music',
    'unlimited_data',
    'paperless_billing',
    'under_30'
]

for col in binary_cols:
    print(col)
    print(data[col].value_counts())
    print("-" * 40)

senior_citizen
senior_citizen
No     5901
Yes    1142
Name: count, dtype: int64
----------------------------------------
married
married
No     3641
Yes    3402
Name: count, dtype: int64
----------------------------------------
dependents
dependents
No     5416
Yes    1627
Name: count, dtype: int64
----------------------------------------
referred_a_friend
referred_a_friend
No     3821
Yes    3222
Name: count, dtype: int64
----------------------------------------
phone_service
phone_service
Yes    6361
No      682
Name: count, dtype: int64
----------------------------------------
multiple_lines
multiple_lines
No     4072
Yes    2971
Name: count, dtype: int64
----------------------------------------
internet_service
internet_service
Yes    5517
No     1526
Name: count, dtype: int64
----------------------------------------
online_security
online_security
No     5024
Yes    2019
Name: count, dtype: int64
----------------------------------------
online_backup
online_backup
No     4614
Yes 

In [9]:
binary_cols = [
    'senior_citizen',
    'married',
    'dependents',
    'referred_a_friend',
    'phone_service',
    'multiple_lines',
    'internet_service',
    'online_security',
    'online_backup',
    'device_protection_plan',
    'premium_tech_support',
    'streaming_tv',
    'streaming_movies',
    'streaming_music',
    'unlimited_data',
    'paperless_billing',
    'under_30'
]

for col in binary_cols:
    data[col] = data[col].map({
        "Yes": 1,
        "No": 0
    })

In [10]:
data[binary_cols].head()

,senior_citizen,married,dependents,referred_a_friend,phone_service,multiple_lines,internet_service,online_security,online_backup,device_protection_plan,premium_tech_support,streaming_tv,streaming_movies,streaming_music,unlimited_data,paperless_billing,under_30
0,1,0,0,0,0,0,1,0,0,1,0,0,1,0,0,1,0
1,1,1,1,1,1,1,1,0,1,0,0,0,0,0,1,1,0
2,1,0,1,0,1,1,1,0,0,0,0,1,1,1,1,1,0
3,1,1,1,1,1,0,1,0,1,1,0,1,1,0,1,1,0
4,1,1,1,1,1,1,1,0,0,0,0,0,0,0,1,1,0


In [11]:
data[binary_cols].isna().sum()

senior_citizen            0
married                   0
dependents                0
referred_a_friend         0
phone_service             0
multiple_lines            0
internet_service          0
online_security           0
online_backup             0
device_protection_plan    0
premium_tech_support      0
streaming_tv              0
streaming_movies          0
streaming_music           0
unlimited_data            0
paperless_billing         0
under_30                  0
dtype: int64

### Binary Encoding

Binary categorical features were converted from text values (Yes/No) into numerical values (1/0).

Encoding was applied to customer demographics, service subscriptions, and account-related features.

This transformation is necessary because machine learning algorithms operate on numerical data and cannot directly interpret categorical text values.

In [12]:
data["churn_label"] = data["churn_label"].map({
    "Yes": 1,
    "No": 0
})

data["churn_label"].value_counts()

churn_label
0    5174
1    1869
Name: count, dtype: int64

### Target Encoding

The target variable (churn_label) was converted into a binary numerical format:

- Yes → 1 (Customer Churned)
- No → 0 (Customer Retained)

This transformation is required because machine learning algorithms operate on numerical values rather than text labels.

In [13]:
data["avg_revenue_per_month"] = (
    data["total_revenue"] /
    data["tenure_in_months"]
)

data[
    [
        "total_revenue",
        "tenure_in_months",
        "avg_revenue_per_month"
    ]
].head()

,total_revenue,tenure_in_months,avg_revenue_per_month
0,59.65,1,59.650000
1,1024.10,8,128.012500
2,1910.88,18,106.160000
3,2995.07,25,119.802800
4,3102.36,37,83.847568


### Feature 1: Average Revenue Per Month

This feature was created by dividing Total Revenue by Tenure in Months.

It represents the average monthly revenue generated from each customer over their entire relationship with the company.

Unlike Monthly Charge, which reflects the current billing amount, this feature captures the long-term financial behavior of the customer.

This helps provide a more stable and historical view of customer value.

In [14]:
contract_risk_map = {
    "Month-to-Month": 3,
    "One Year": 2,
    "Two Year": 1
}

data["contract_risk_score"] = data["contract"].map(contract_risk_map)

data[["contract", "contract_risk_score"]].head()

,contract,contract_risk_score
0,Month-to-Month,3
1,Month-to-Month,3
2,Month-to-Month,3
3,Month-to-Month,3
4,Month-to-Month,3


In [15]:
data["contract"].value_counts()

contract
Month-to-Month    3610
Two Year          1883
One Year          1550
Name: count, dtype: int64

### Feature 2: Contract Risk Score

The contract type was transformed into a numerical risk score based on churn behavior observed during EDA.

- Month-to-Month → High Risk (3)
- One Year → Medium Risk (2)
- Two Year → Low Risk (1)

This encoding introduces an ordinal relationship between contract types, capturing customer commitment level more effectively than one-hot encoding.

In [16]:
service_cols = [
    "online_security",
    "online_backup",
    "device_protection_plan",
    "premium_tech_support",
    "streaming_tv",
    "streaming_movies",
    "streaming_music",
    "multiple_lines",
    "phone_service",
    "internet_service",
    "unlimited_data"
]

data["service_engagement_score"] = data[service_cols].sum(axis=1)

In [17]:
data[["service_engagement_score"] + service_cols].head()

,service_engagement_score,online_security,online_backup,device_protection_plan,premium_tech_support,streaming_tv,streaming_movies,streaming_music,multiple_lines,phone_service,internet_service,unlimited_data
0,3,0,0,1,0,0,1,0,0,0,1,0
1,5,0,1,0,0,0,0,0,1,1,1,1
2,7,0,0,0,0,1,1,1,1,1,1,1
3,7,0,1,1,0,1,1,0,0,1,1,1
4,4,0,0,0,0,0,0,0,1,1,1,1


### Feature 3: Service Engagement Score

Since binary encoding was already applied (Yes/No → 1/0), the feature was created by directly summing service-related columns.

This feature represents the total number of services used by each customer.

Higher values indicate stronger engagement with the company’s services, while lower values suggest weaker engagement and higher churn risk.

In [18]:
data["referral_engagement_score"] = (
    data["referred_a_friend"] + data["number_of_referrals"]
)

In [19]:
data[["referred_a_friend", "number_of_referrals", "referral_engagement_score"]].head()

,referred_a_friend,number_of_referrals,referral_engagement_score
0,0,0,0
1,1,1,2
2,0,0,0
3,1,1,2
4,1,1,2


### Feature 4: Referral Engagement Score

This feature combines two variables:
- Whether the customer referred a friend
- The total number of referrals made

The goal is to measure customer loyalty and engagement through word-of-mouth behavior.

Higher values indicate highly satisfied and loyal customers who actively promote the service, while lower values indicate weak or no engagement in referrals.

In [20]:
data["value_stability_score"] = (
    data["avg_revenue_per_month"] - data["monthly_charge"]
)

In [21]:
data[[
    "monthly_charge",
    "avg_revenue_per_month",
    "value_stability_score"
]].head()

,monthly_charge,avg_revenue_per_month,value_stability_score
0,39.65,59.650000,20.000000
1,80.65,128.012500,47.362500
2,95.45,106.160000,10.710000
3,98.50,119.802800,21.302800
4,76.50,83.847568,7.347568


### Feature 5: Value Stability Score

This feature measures the difference between a customer's current monthly charge and their historical average monthly revenue.

It helps capture changes in customer spending behavior over time, which is a key indicator of churn risk.

- Positive values may indicate declining customer value.
- Negative values may indicate growth in customer spending.
- Values close to zero indicate stable customer behavior.

In [22]:
service_groups = {
    "security": ["online_security", "online_backup", "device_protection_plan", "premium_tech_support"],
    "streaming": ["streaming_tv", "streaming_movies", "streaming_music"],
    "connectivity": ["phone_service", "multiple_lines", "internet_service", "unlimited_data"]
}

data["service_breadth_score"] = 0

for group, cols in service_groups.items():
    data[group] = data[cols].sum(axis=1)
    data["service_breadth_score"] += (data[group] > 0).astype(int)

In [23]:
data[[
    "security",
    "streaming",
    "connectivity",
    "service_breadth_score"
]].head()

,security,streaming,connectivity,service_breadth_score
0,1,1,1,3
1,1,0,4,2
2,0,3,4,2
3,2,2,3,3
4,0,0,4,1


### Feature 6: Service Breadth Score

This feature measures how diversified a customer's service usage is across different categories (security, streaming, and connectivity).

Unlike engagement score, which measures quantity, this feature captures variety in usage.

Higher diversity indicates stronger customer integration with the company ecosystem and lower churn risk.

In [24]:
new_features = [
    "avg_revenue_per_month",
    "contract_risk_score",
    "service_engagement_score",
    "referral_engagement_score",
    "value_stability_score",
    "service_breadth_score"
]

data[new_features].describe().round(2)

,avg_revenue_per_month,contract_risk_score,service_engagement_score,referral_engagement_score,value_stability_score,service_breadth_score
count,7043.00,7043.00,7043.00,7043.00,7043.00,7043.00
mean,88.12,2.25,5.17,2.41,23.36,2.11
std,35.91,0.85,2.89,3.37,15.96,0.85
min,18.17,1.00,1.00,0.00,-13.56,1.00
25%,56.77,1.00,3.00,0.00,9.49,1.00
50%,93.25,3.00,5.00,0.00,23.20,2.00
75%,117.02,3.00,7.00,4.00,36.70,3.00
max,190.70,3.00,11.00,12.00,121.20,3.00


In [25]:
corr_matrix = data.select_dtypes(
    include=["int64", "float64"]
).corr()

corr_matrix["churn_label"].sort_values(
    key=abs,
    ascending=False
)

churn_label                          1.000000
satisfaction_score                  -0.754649
contract_risk_score                  0.435398
tenure_in_months                    -0.352861
number_of_referrals                 -0.286540
referral_engagement_score           -0.277025
dependents                          -0.248542
internet_service                     0.227890
total_long_distance_charges         -0.223756
total_revenue                       -0.223003
number_of_dependents                -0.218780
total_charges                       -0.198546
monthly_charge                       0.193356
paperless_billing                    0.191825
connectivity                         0.179927
security                            -0.173061
online_security                     -0.171226
avg_revenue_per_month                0.170906
unlimited_data                       0.166545
premium_tech_support                -0.164674
senior_citizen                       0.150889
married                           

## Feature Engineering Summary

### Retained Features

The following engineered features were retained because they demonstrated meaningful relationships with customer churn and added business value beyond the original variables:

- **Avg Revenue Per Month**
  - Measures the average revenue generated by a customer per month.
  - Helps capture customer value independently of tenure length.

- **Contract Risk Score**
  - Converts contract types into a risk-based scale.
  - Customers with Month-to-Month contracts are assigned higher risk scores than customers with long-term contracts.

- **Referral Engagement Score**
  - Combines referral behavior and the number of referrals made by a customer.
  - Represents customer engagement and loyalty to the company.

---

### Removed Features

The following engineered features were removed because they showed very weak relationships with churn and did not provide sufficient predictive value:

- **Service Engagement Score**
- **Service Breadth Score**
- **Value Stability Score**

Removing these features helps reduce unnecessary complexity while improving model interpretability.

---

### Key Takeaway

Feature engineering should focus on creating variables that capture meaningful customer behavior and business patterns.

After evaluating the engineered features, only the features that demonstrated useful relationships with churn were retained for modeling.

In [26]:
features_to_drop = [
    "service_engagement_score",
    "service_breadth_score",
    "value_stability_score"
]

data = data.drop(columns=features_to_drop)

In [27]:
data.select_dtypes(include="object").columns.tolist()

['city', 'offer', 'internet_type', 'contract', 'payment_method']

In [28]:
data["city"].nunique()

1106

In [29]:
data = data.drop(columns=["city"])

## Categorical Encoding

The remaining categorical features contain multiple categories and cannot be represented using binary encoding.

One-Hot Encoding will be applied to:

- Contract
- Payment Method
- Internet Type
- Offer

The City feature will be removed because it contains a very high number of unique categories (high cardinality), which could introduce unnecessary complexity and increase the dimensionality of the dataset.

In [30]:
categorical_cols = [
    "contract",
    "payment_method",
    "internet_type",
    "offer"
]

data = pd.get_dummies(
    data,
    columns=categorical_cols,
    drop_first=True,
    dtype=int
)

In [31]:
data.shape

(7043, 54)

In [32]:
data.head()

,gender,age,under_30,senior_citizen,married,dependents,number_of_dependents,zip_code,latitude,longitude,population,referred_a_friend,number_of_referrals,tenure_in_months,phone_service,avg_monthly_long_distance_charges,multiple_lines,internet_service,avg_monthly_gb_download,online_security,online_backup,device_protection_plan,premium_tech_support,streaming_tv,streaming_movies,streaming_music,unlimited_data,paperless_billing,monthly_charge,total_charges,total_refunds,total_extra_data_charges,total_long_distance_charges,total_revenue,satisfaction_score,churn_label,avg_revenue_per_month,contract_risk_score,referral_engagement_score,security,streaming,connectivity,contract_One Year,contract_Two Year,payment_method_Credit Card,payment_method_Mailed Check,internet_type_DSL,internet_type_Fiber Optic,internet_type_No Internet,offer_Offer A,offer_Offer B,offer_Offer C,offer_Offer D,offer_Offer E
0,1,78,0,1,0,0,0,90022,34.023810,-118.156582,68701,0,0,1,0,0.00,0,1,8,0,0,1,0,0,1,0,0,1,39.65,39.65,0.00,20,0.00,59.65,3,1,59.650000,3,0,1,1,1,0,0,0,0,1,0,0,0,0,0,0,0
1,0,74,0,1,1,1,1,90063,34.044271,-118.185237,55668,1,1,8,1,48.85,1,1,17,0,1,0,0,0,0,0,1,1,80.65,633.30,0.00,0,390.80,1024.10,3,1,128.012500,3,2,1,0,4,0,0,1,0,0,1,0,0,0,0,0,1
2,1,71,0,1,0,1,3,90065,34.108833,-118.229715,47534,0,0,18,1,11.33,1,1,52,0,0,0,0,1,1,1,1,1,95.45,1752.55,45.61,0,203.94,1910.88,2,1,106.160000,3,0,0,3,4,0,0,0,0,0,1,0,0,0,0,1,0
3,0,78,0,1,1,1,1,90303,33.936291,-118.332639,27778,1,1,25,1,19.76,0,1,12,0,1,1,0,1,1,0,1,1,98.50,2514.50,13.43,0,494.00,2995.07,2,1,119.802800,3,2,2,2,3,0,0,0,0,0,1,0,0,0,1,0,0
4,0,80,0,1,1,1,1,90602,33.972119,-118.020188,26265,1,1,37,1,6.33,1,1,14,0,0,0,0,0,0,0,1,1,76.50,2868.15,0.00,0,234.21,3102.36,2,1,83.847568,3,2,0,0,4,0,0,0,0,0,1,0,0,0,1,0,0


In [33]:
corr_matrix["churn_label"][
    ["security", "streaming", "connectivity"]
]

security       -0.173061
streaming       0.065970
connectivity    0.179927
Name: churn_label, dtype: float64

In [34]:
data.isna().sum().sum()

np.int64(0)

In [35]:
data.dtypes.value_counts()

int64      45
float64     9
Name: count, dtype: int64

In [36]:
data["churn_label"].value_counts()

churn_label
0    5174
1    1869
Name: count, dtype: int64

In [37]:
data.to_csv(
    "telco_feature_engineered.csv",
    index=False
)

# Final Dataset Summary

## Dataset Information

- Rows: 7,043
- Features engineered and evaluated
- Categorical variables encoded
- No missing values
- Target variable encoded
- Dataset fully numerical and ready for machine learning

## Feature Engineering Highlights

Retained Features:

- Avg Revenue Per Month
- Contract Risk Score
- Referral Engagement Score

Removed Features:

- Service Engagement Score
- Service Breadth Score
- Value Stability Score

## Encoding Applied

- Binary Encoding for Yes/No variables
- Label Encoding for Gender
- Target Encoding for Churn Label
- One-Hot Encoding for multi-category features

## Final Status

The dataset is now fully prepared for model development, training, and evaluation.